v1/topdown/eval.go

```diff
@@ -2014,8 +2135,8 @@ func (e evalFunc) evalOneRule(iter unifyIterator, rule *ast.Rule, cacheKey ast.R
 			}
 
 			if len(rule.Head.Args) == len(e.terms)-1 {
-				if result.Value.Compare(ast.Boolean(false)) == 0 {
-					if prev != nil && ast.Compare(prev, result) != 0 {
+				if ast.Boolean(false).Equal(result.Value) {
+					if prev != nil && !prev.Equal(result) {
 						return functionConflictErr(rule.Location)
 					}
 					prev = result
```

Comment:
OK so this could be be a problem, one that I ran into before:

Before, `prev` was passed in. If this function is executed twice, the second call would *"see"* the `prev` assignment done by the first call.

Now, the first call is passed `nil`. Whatever it does assign to `prev` would *not* be "seen" by the second call.

If it's a problem, we'll notice that function value conflicts disappear, like

```rego
x if f(1) == 2

f(_) := 2
f(_) := 3 # should give a conflict error when `f(1)` is evaluated
```

Reply:
Interesting! Would the pointer not have to be dereferenced for the assignment to be effectual outside of the function?

IIRC, passing nil instead of `prev` was something I did just for readability ("it's always nil here, so better show that"), but that's hardly a hill I'll choose to die on. I'll update the code to pass `prev` as before 👍 Although it certainly sounds like a unit test or two has gone missing if that's not covered?

Reply:
Yeah I've been wondering about that, too. In the end, tests trump everything -- perhaps my recollection above is just wrong.